Implement rest-xml codegen

adimarco opened the pull request:
This is still a work in progress.  XML serialization still needs to be done for request bodies that contain XML, but requests without it and responses with both headers and XML in the body are looking good.

I'm using S3 as the test case.  Now that HTTP dispatch is completely abstracted behind the `DispatchSignedRequest` trait, I've implemented a `MockRequestDispatcher` that can be used to test the generated service with the sample requests and responses provided in the AWS documentation.  There's the beginnings of that in `src/s3.rs` - see [here](https://github.com/rusoto/rusoto/blob/9c9156a971dfe953a0a91922b2e65ea07818a589/src/s3.rs#L14) for an example.  I think this is a really good way to do effective unit tests of the generated services without actually talking to Amazon, and I intend to use it everywhere.

There's still a few more days worth of work here, but I have this in mind as the centerpiece of the next release.  Hopefully this weekend.

This technically implements the `head_object()` functionality that @ammojamo was looking for in #18 - but I'd wait until it's a little more stable until using it unless you're desperate :smile:

Note:  This builds on #310, so that will need to get merged first.

matthewkmayer commented:
Whoever picks this up please follow changes in https://github.com/rusoto/rusoto/pull/393 and https://github.com/rusoto/rusoto/pull/391 to not generate unneeded serializers and deserializers.

indiv0 commented:
@adimarco anyway I can help with this PR?

adimarco commented:
I'm actually a few hours worth of focused work away from finishing this up, at least enough to remove the "WIP" and ask for some real review.  I just haven't had any time at all in the last few weeks to put into it.  I'll make a point to get into it this weekend and see if I can wrap it up.

adimarco commented:
Some pretty exhaustive integration tests of the generated S3 code are working nicely now, and I'm ready for some actual review of this pull request to begin.  See `tests/s3.rs` for those integration tests.

The bulk of the work is in `codegen/src/generator/rest_xml.rs`

matthewkmayer commented:
@adimarco is this ready for review?  Can you rebase and squash the branch?

adimarco commented:
Yeah, this is ready for some feedback at a bare minimum.  It's generating an S3 client that works.  I'll add another bunch of unit and integration tests before it actually gets merged, but it's a big review and it would be good to start getting some eyes on it now for anything crazy I may have done.  I did this pull review in a bunch of disconnected sessions over several months and I may have missed something somewhere.

I'll get it rebased and squashed.

adimarco commented on helpers/tests/s3.rs:
```diff
@@ -0,0 +1,96 @@
+#![cfg(feature = "s3")]
```
I need to yank this whole file out.  I moved the S3Helper when I started the rewrite, but forgot it was there.  This almost certainly doesn't compile or work.

indiv0 replied:
I think we should find a way to maintain the tests, at least, in case this PR added a regression.

adimarco commented on src/param.rs:
```diff
@@ -5,15 +5,21 @@
 
 use std::collections::BTreeMap;
-pub type Params = BTreeMap<String, String>;
+pub type Params = BTreeMap<String, Option<String>>;
```
This is necessary because S3 has some URLs in the format `/foo?bar` where the key in the query string has no associated value.

adimarco commented on tests/s3.rs:
```diff
@@ -1,96 +1,165 @@
 #![cfg(feature = "s3")]
```
These integration tests were a total rewrite, no need to compare to the old on the left, it's all new.

indiv0 commented on Cargo.toml:
```diff
@@ -38,6 +38,7 @@ serde_json = "0.8.0"
 time = "0.1.35"
 url = "1.2.0"
 xml-rs = "0.1.26"
+data-encoding = "1.1.2"
```
Non-alphabetical dependencies... :/

indiv0 commented on codegen/src/generator/mod.rs:
```diff
@@ -107,15 +110,15 @@ where P: GenerateProtocol {
-            &Some(ref service_abbreviation) => service_abbreviation.as_str(),
-            &None => service.metadata.service_full_name.as_ref()
+        &Some(ref service_abbreviation) => service_abbreviation.as_str(),
+        &None => service.metadata.service_full_name.as_ref()
```
Why have these lines been un-indented?

indiv0 commented on codegen/src/generator/mod.rs:
```diff
@@ -143,33 +146,55 @@ fn generate_primitive_type(name: &str, shape_type: ShapeType, for_timestamps: &s
+        _ => capitalized
+    }
+
+}
```
Why all the double newlines in this function?

indiv0 commented on codegen/src/generator/mod.rs:
```diff
@@ -262,11 +287,8 @@ fn generate_struct_fields<P>(service: &Service, shape: &Shape, shape_name: &str,
-            _ => format!("{}Error", self.name)
-        }
+        let type_name = mutate_type_name(&self.name);
+        format!("{}Error", type_name)
```
If I understand correctly, one of the type names S3 provides is `Error`, which is turned into `S3Error` via `mutate_type_name`. Does this function return `S3ErrorError` then? Or is `S3Error `never used with this function?

adimarco replied:
The latter.  The only place `S3Error` is actually used at all is to define `pub type Errors = Vec<S3Error>;`, and that's the type that gets used everywhere `error_type_name()` is called, so it doesn't end up duplicated as `S3ErrorError`

indiv0 commented on codegen/src/generator/rest_xml.rs:
```diff
@@ -0,0 +1,855 @@
+    Some(shape.members.as_ref().unwrap().iter().filter_map(|(member_name, member)| {
+        if member.location.is_none() {
+            return None;
+        }
```
This check chould be merged into the below `match` statement. But this way is fine too, I guess.

adimarco replied:
Ooh, match guards.  There's a bit of new syntax for me to file away for future use.  `Some(location) if location == "querystring"`

indiv0 commented on codegen/src/generator/rest_xml.rs:
```diff
@@ -0,0 +1,855 @@
+fn generate_list_deserializer(shape: &Shape) -> String {
+
+    // flattened lists are just the list elements repeated without
+    // an enclusing <FooList></FooList> tag
```
Typo: `enclusing`

indiv0 commented on codegen/src/generator/rest_xml.rs:
```diff
@@ -0,0 +1,855 @@
+        return generate_flat_list_deserializer(shape);
+    }
+
+    let location_name = shape.member.as_ref().and_then(|m| m.location_name.as_ref()).map(|name| &name[..]).unwrap_or(shape.member());
```
Couldn't the `and_then` and `map` be joined?

indiv0 commented on codegen/src/generator/rest_xml.rs:
```diff
@@ -0,0 +1,855 @@
+
+}
+
+fn generate_map_serializer(_shape: &Shape) -> String {
```
Is this function named correctly?

adimarco replied:
Actually, that function is completely unimplemented.  Good catch.  This means that S3 doesn't need to serialize Map types in request bodies (or at least not for any of the methods I've tested), but I'd be willing to bet that this will be needed by _something_ with a rest-xml protocol.  I'll implement that before this gets merged.

adimarco commented:
I'll run the whole `rest_xml` generator through rustfmt.  A combination of about 4 different editing environments and varying states of awake over the course of the PR are to blame for weird whitespace.

Adding the existing S3 tests to the new build is the only real chunk of work that's left here.  I'll get on that and the handful of other things in the review later today.  Glad to hear it looks sane.  I've been staring at this so long it all sort of blurs together :)

adimarco commented:
Rebased on top of the recent changes from master and implemented @indiv0 's code review feedback.

Unless there's objection I'm going to just file an issue for the `serialize_map` method being unimplemented, because we don't need it for S3 and I really want to get this out.  I can address that when the rest of the `rest-xml` services are added with this code generator, since it'll invariably involve a handful of other tweaks.

I'll get the unit tests from `master` translated into something that makes sense in this context and get them committed, and then we can get this thing merged.

matthewkmayer commented:
@adimarco is translating unit tests the only thing left on this PR?

matthewkmayer commented on src/signature.rs:
```diff
@@ -311,13 +311,17 @@ fn build_canonical_query_string(params: &Params) -> String {
-        output.push_str(&encode_uri_strict(item.1));
+
+        output.push_str(&encode_uri_strict(key));
+        output.push_str("=");        
```
Should this `=` be inside the `if` statement just below?  So a url can have `foo?bar` instead of `foo?bar=` if there's no value?

matthewkmayer commented on Cargo.toml:
```diff
@@ -26,6 +26,7 @@ version = ">= 0.4.0"
 
 [dependencies]
 chrono = "0.2.21"
+data-encoding = "1.1.2"
```
Can we use our existing base64 encoding/decoding dependency, rustc_serialize?

matthewkmayer commented:
Things I'd like addressed, but don't need to be done before merging:

* extra newlines abound
* bringing in a new library when we have rustc_serialize
* the `=` or not question for query string params without values
* fix up the test by printing the results instead of just unwrapping

I'd rather "fix those in post" and land this than wait much longer on this PR.  Six months in flight!

matthewkmayer commented:
@adimarco concurs, time to :shipit: .

adimarco commented:
Opened #500 to that end
